ssl/statem/statem_clnt.c

```diff
@@ -3089,14 +3092,26 @@ static int tls_construct_cke_dhe(SSL *s, WPACKET *pkt)
     }
 
     /* send off the data */
+    DH_get0_pqg(dh_clnt, &p, NULL, NULL);
     DH_get0_key(dh_clnt, &pub_key, NULL);
-    if (!WPACKET_sub_allocate_bytes_u16(pkt, BN_num_bytes(pub_key),
+    /*
+     * for interoperability with some versions of the Microsoft TLS
+     * stack, we need to zero pad the DHE pub key to the same length
+     * as the prime, so use the length of the prime here
+     */
+    if (!WPACKET_sub_allocate_bytes_u16(pkt, BN_num_bytes(p),
                                         &keybytes)) {
         SSLfatal(s, SSL_AD_INTERNAL_ERROR, SSL_F_TLS_CONSTRUCT_CKE_DHE,
                  ERR_R_INTERNAL_ERROR);
         goto err;
     }
 
+    npad = (BN_num_bytes(p) - BN_num_bytes(pub_key));
+    for(j = 0; j < npad; ++j) {
+        *keybytes = 0;
+        ++keybytes;
+    }
+
     BN_bn2bin(pub_key, keybytes);
```

Comment:
You may want to look at `BN_bn2binpad`.

Reply:
Thank you, good idea! I need to close and reopen the pull request because of cla-check.